scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetReadFn.scala

```diff
@@ -0,0 +1,237 @@
+import org.apache.beam.sdk.transforms.splittabledofn.{OffsetRangeTracker, RestrictionTracker}
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableSet
+import org.apache.parquet.filter2.compat.FilterCompat
+import org.apache.parquet.hadoop.ParquetFileReader
+import org.apache.parquet.hadoop.api.InitContext
+import org.apache.parquet.hadoop.metadata.BlockMetaData
+import org.apache.parquet.io.{ColumnIOFactory, ParquetDecodingException, RecordReader}
+import org.apache.parquet.{HadoopReadOptions, ParquetReadOptions}
+import org.slf4j.LoggerFactory
+
+import java.util.{Set => JSet}
+import scala.jdk.CollectionConverters._
+import scala.util.Try
+
+class ParquetReadFn[T, R](
+  readSupportFactory: ReadSupportFactory[T],
+  conf: SerializableConfiguration,
+  projectionFn: SerializableFunction[T, R]
+) extends DoFn[ReadableFile, R] {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+  private val SPLIT_LIMIT = 64000000L
+
+  private def parquetFileReader(file: ReadableFile): ParquetFileReader = {
+    val options = HadoopReadOptions.builder(conf.get()).build
+    ParquetFileReader.open(BeamInputFile.of(file.openSeekable), options)
+  }
+
+  @GetRestrictionCoder def getRestrictionCoder = new OffsetRange.Coder
+
+  @GetInitialRestriction
+  def getInitialRestriction(@Element file: ReadableFile): OffsetRange = {
+    val reader = parquetFileReader(file)
+
+    val offsetRangeResult = Try(new OffsetRange(0, reader.getRowGroups.size)).toEither
+    reader.close()
+
+    offsetRangeResult.fold(throw _, identity)
+  }
+
+  @NewTracker def newTracker(@Restriction restriction: OffsetRange) =
+    new OffsetRangeTracker(restriction)
+
+  @GetSize
+  def getSize(@Element file: ReadableFile, @Restriction restriction: OffsetRange): Double = {
+    val (_, size) = getRecordCountAndSize(file, restriction)
+    size
+  }
+
+  @SplitRestriction
+  def split(
+    @Restriction restriction: OffsetRange,
+    output: DoFn.OutputReceiver[OffsetRange],
+    @Element file: ReadableFile
+  ): Unit = {
+    val reader = parquetFileReader(file)
+
+    val splitResult = Try {
+      splitRowGroupsWithLimit(
+        restriction.getFrom,
+        restriction.getTo,
+        reader.getRowGroups.asScala.toSeq
+      ).foreach(output.output)
+    }.toEither
+
+    reader.close()
+
+    splitResult.fold(throw _, identity)
+  }
+
+  @ProcessElement
+  def processElement(
+    @Element file: ReadableFile,
+    tracker: RestrictionTracker[OffsetRange, Long],
+    outputReceiver: DoFn.OutputReceiver[R]
+  ): Unit = {
+    logger.debug(
+      "start {} to {}",
+      tracker.currentRestriction.getFrom,
+      tracker.currentRestriction.getTo
+    )
+    val options: ParquetReadOptions = HadoopReadOptions.builder(conf.get()).build
+
+    val reader = ParquetFileReader.open(BeamInputFile.of(file.openSeekable), options)
+    try {
+      val filter = options.getRecordFilter
+      val hadoopConf = options.asInstanceOf[HadoopReadOptions].getConf
```

Comment:
Why not use `as` ?
```suggestion
      val hadoopConf = options.as(classOf[HadoopReadOptions]).getConf
```

Reply:
ah, that's a different type of options - [https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java)